Sources/MongoSwift/BSON/Document.swift

```diff
@@ -157,17 +157,51 @@ extension Document {
 
                 // otherwise, we just create a new document and replace this key
             } else {
-                // TODO: SWIFT-224: use va_list variant of bson_copy_to_excluding to improve performance
-                var newSelf = Document()
+                guard let iter = DocumentIterator(forDocument: self) else {
+                    throw InternalError(message: "BSON buffer is unexpectedly too small (< 5 bytes)")
+                }
+
+                var keysBefore: [String] = []
+                var keysAfter: [String] = []
                 var seen = false
-                try self.forEach { pair in
-                    if !seen && pair.key == key {
+
+                while let next = iter.next() {
+                    if next.key == key {
                         seen = true
-                        try newSelf.setValue(for: pair.key, to: newValue)
                     } else {
-                        try newSelf.setValue(for: pair.key, to: pair.value)
+                        seen ? keysAfter.append(next.key) : keysBefore.append(next.key)
+                    }
+                }
+
+                // To preserve order, we construct a Document excluding keys before the key and a Document excluding
+                // keys after the key. We then merge both Documents and appropriately set the new value for the key.
+                guard let bson = bson_new() else {
+                    throw InternalError(message: "Invalid BSON data")
```

Comment:
we might consider making this error a static constant so we can reuse it any place we do this. here's an example of another one of those in the driver: https://github.com/mongodb/mongo-swift-driver/blob/33aed8834adb1c07d811b3431ceceee1355784a4/Sources/MongoSwift/ClientSession.swift#L36

Reply:
Done.

Reply:
Done.